### PR TITLE
New Resource: aws_servicequotas_service_quota

### LINKED ISSUE #9118: Support Managing Service Quotas
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

The new Service Quotas service provides a simplified method for requesting quota (limit) increases. Theoretically, Terraform can support this as a resource by submitting the request via the new API and reading back either request values and/or current service quotas (seems preferable!).

A separate feature request is created for managing Service Quotas Quota Request Templates and their associations for Organizations: #9119 

### New Resource(s)

* `aws_servicequotas_service_quota`
* `aws_servicequotas_service_quota_increase_request`

#### Option 1: Increase request managing resource

Simply submits quota increases. No lifecycle management for service quotas if they are increased in the future.

```hcl
# Relevant API Calls:
#   GetRequestedServiceQuotaChange
#   RequestServiceQuotaIncrease
resource "aws_servicequotas_service_quota_increase_request" "example" {
  desired_value = 123   # TypeFloat, Required, ForceNew
  quota_code    = "..." # TypeString, Required, ForceNew
  service_code  = "..." # TypeString, Required, ForceNew
}
```

#### Option 2: Service quota managing resource

Manages lifecycle of service quota. Requests quota increase if new value is greater than current value. Value is based on desired value of quota increase request if pending or case opened state, otherwise based on current service quota.

```hcl
# Relevant API Calls:
#   GetRequestedServiceQuotaChange
#   GetServiceQuota
#   RequestServiceQuotaIncrease
# Additional attributes:
#   request_id # TypeString, Computed
#   request_status # TypeString, Computed
resource "aws_servicequotas_service_quota" "example" {
  quota_code    = "..." # TypeString, Required, ForceNew
  service_code  = "..." # TypeString, Required, ForceNew
  value         = 123   # TypeFloat, Required
}
```

### References

* https://aws.amazon.com/about-aws/whats-new/2019/06/introducing-service-quotas-view-and-manage-quotas-for-aws-services-from-one-location/
* https://aws.amazon.com/blogs/mt/introducing-service-quotas-view-and-manage-your-quotas-for-aws-services-from-one-central-location/
* `GetRequestedServiceQuotaChange`
* `GetServiceQuota`
* `RequestServiceQuotaIncrease`

### PR DESCRIPTION
**Pull request by bflad**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" comments, they generate extra noise for pull request followers and do not help prioritize the request

Replacing #9121 due to bad rebase...
Closes #9118

Release note for [CHANGELOG](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md):

```release-note
* **New Resource:** `aws_servicequotas_service_quota`
```

Output from acceptance testing (requesting increase of t3.nano and t3.micro limits in testing account):

```
export SERVICEQUOTAS_INCREASE_ON_CREATE_QUOTA_CODE=L-55EB784E
export SERVICEQUOTAS_INCREASE_ON_CREATE_SERVICE_CODE=ec2
export SERVICEQUOTAS_INCREASE_ON_CREATE_VALUE=25
export SERVICEQUOTAS_INCREASE_ON_UPDATE_QUOTA_CODE=L-0F0ED007
export SERVICEQUOTAS_INCREASE_ON_UPDATE_SERVICE_CODE=ec2
export SERVICEQUOTAS_INCREASE_ON_UPDATE_VALUE=25

--- PASS: TestAccAwsServiceQuotasServiceQuota_basic (13.83s)
--- PASS: TestAccAwsServiceQuotasServiceQuota_Value_IncreaseOnCreate (12.52s)
--- PASS: TestAccAwsServiceQuotasServiceQuota_Value_IncreaseOnUpdate (20.00s)
```

Output from acceptance testing in AWS GovCloud (US):

```
--- SKIP: TestAccAwsServiceQuotasServiceQuota_basic (2.51s)
    resource_aws_servicequotas_service_quota_test.go:138: skipping acceptance testing: RequestError: send request failed
        caused by: Post https://servicequotas.us-gov-west-1.amazonaws.com/: dial tcp: lookup servicequotas.us-gov-west-1.amazonaws.com: no such host
```

### CODE REVIEW

File: aws/provider.go [hunk 1]
```diff
@@ -681,6 +681,7 @@ func Provider() terraform.ResourceProvider {
 			"aws_service_discovery_private_dns_namespace":             resourceAwsServiceDiscoveryPrivateDnsNamespace(),
 			"aws_service_discovery_public_dns_namespace":              resourceAwsServiceDiscoveryPublicDnsNamespace(),
 			"aws_service_discovery_service":                           resourceAwsServiceDiscoveryService(),
+			"aws_servicequotas_service_quota":                         resourceAwsServiceQuotasServiceQuota(),
```
Comment (nywilken) [inline aws/provider.go]: I know we generally use the AWS service name for the resource name, but this name has a bit of a stutter. What do you about just using `aws_service_quotas`?

Comment (bflad) [reply aws/provider.go]: This is noted in the [Contributing Guide](https://github.com/terraform-providers/terraform-provider-aws/blob/master/.github/CONTRIBUTING.md#new-resource) with the following item:

> *Naming:* Resources should be named `aws_<service>_<name>` where `service` is the AWS short service name and `name` is a short, preferably single word, description of the resource. Use `_` as a separator.

What isn't included in the guide is the reasoning: 

* It helps operators and maintainers easier find resources by AWS service or configure [Customized Endpoints](https://www.terraform.io/docs/providers/aws/guides/custom-service-endpoints.html#available-endpoint-customizations) using the service name that is part of the resource name. There will be other resources with this service: https://github.com/terraform-providers/terraform-provider-aws/issues/9119
* Prevents conflicting naming issues should AWS release other "quota" related services/resources in the future
* Potentially sets up the resources for easier auto-generation in the future

We likely should include some/all of this rationale either in the contributing or maintaining guides for easier reference. I'll raise a PR for the contributing guide to start.

Comment (bflad) [reply aws/provider.go]: See also: #9347 👍

Comment (nywilken) [reply aws/provider.go]: That makes sense. Thanks for rationale. I found the existing data sources after leaving the comment and quickly thought to rescind this comment. Especially, after thinking more about "Prevents conflicting naming issues should AWS release other "quota" related services/resources in the future". The contributing guide update looks good :+1:

File: aws/resource_aws_servicequotas_service_quota.go [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+
+	if err != nil {
+		return fmt.Errorf("error getting Service Quotas Service Quota (%s): %s", d.Id(), err)
+	}
+
+	if output == nil {
+		return fmt.Errorf("error getting Service Quotas Service Quota (%s): empty result", d.Id())
+	}
+
+	if value > aws.Float64Value(output.Quota.Value) {
```
Comment (nywilken) [inline aws/resource_aws_servicequotas_service_quota.go]: Providing a value less than the current quota value would result in a noop and exit successfully, correct? If so should there be an error message to the operator that the provided value is less than the currently set quota?

Comment (bflad) [reply aws/resource_aws_servicequotas_service_quota.go]: Correct. Terraform should show automatically show perpetual difference in this case. 😄

File: aws/resource_aws_servicequotas_service_quota.go [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+		d.Set("request_id", output.RequestedQuota.Id)
+	}
+
+	return resourceAwsServiceQuotasServiceQuotaRead(d, meta)
+}
+
+func resourceAwsServiceQuotasServiceQuotaRead(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).servicequotasconn
+
+	requestID := d.Get("request_id").(string)
```
[the remark this answers is not in the document]
Comment (bflad) [reply aws/resource_aws_servicequotas_service_quota.go]: Moving 👍

**Pull request merged**